onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -1024,17 +1024,95 @@ Status TransposeOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, co
 class ReshapeOpBuilder : public BaseOpBuilder {
  public:
   void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) override;
+  static Status AddReshapeOperator(ModelBuilder& model_builder, const Node& node,
+                                   const std::string& input, const std::vector<int32_t>& shape) ORT_MUST_USE_RESULT;
 
  private:
   bool IsOpSupportedImpl(ModelBuilder& model_builder, const Node& node) override;
-
   Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node) override ORT_MUST_USE_RESULT;
+  static bool CanSkipReshape(const Node& node, size_t input_rank, size_t output_rank);
 };
 
 void ReshapeOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) {
   model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());
 }
 
+// We can skip the Reshape if
+// 1. The output of the reshape/flatten is the input 0 of the GEMM/Matmul,
+//    and the input rank >= 2 and output_rand == 2
+//    This is because Gemm/Matmul will map to ANEURALNETWORKS_FULLY_CONNECTED in NNAPI,
+//    ANEURALNETWORKS_FULLY_CONNECTED will flatten the 2+ dim input 0 to 2d
+// 2. Or the output the reshape/flatten is the ouput of the graph
+//    (no op in the graph is using the output except can be used by Gemm/Matmul satisfying condition 1 above)
```

Comment:
nit: rand -> rank ('output_rand' here and 'rand' on line 1068), ouput -> output.

What's checking #2? There's no edge for a graph output so I'd expect you need to check if the output name is found in the graph outputs. Do you also need to check there's only one output edge?

Reply:
If a node has not output edge, can we assume the output is a graph output? Is there any other case can cause this to happen?